### PR TITLE
Bug/quad 237

### PR DESCRIPTION
**Pull request by nischalsamji**
Refactor backend code to add mapper and validators

### CODE REVIEW

File: Quadriga/src/main/java/edu/asu/spring/quadriga/service/publicwebsite/mapper/IAboutTextMapper.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+     */
+    public IAboutText aboutTextDTOtoBean(AboutTextDTO abtDTO);
+
+    /**
+     * This method takes in {@link AboutTextDTO} as an argument and returns a
+     * mapped {@link IAboutText} Object.
+     * 
+     * @param abtText
+     *            Domain Object that needs to be mapped to a
+     *            {@link AboutTextDTO} domain object.
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/service/publicwebsite/mapper/IAboutTextMapper.java]: DTOs are not domain objects

File: Quadriga/src/main/java/edu/asu/spring/quadriga/validator/AboutTextValidator.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
+package edu.asu.spring.quadriga.validator;
+
+import org.springframework.stereotype.Service;
+import org.springframework.validation.Errors;
+import org.springframework.validation.ValidationUtils;
+import org.springframework.validation.Validator;
+
+import edu.asu.spring.quadriga.domain.settings.impl.AboutText;
+
+@Service
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/validator/AboutTextValidator.java]: annotation should be right before class definition

File: Quadriga/src/main/java/edu/asu/spring/quadriga/validator/AboutTextValidator.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ * <code>publicWebsiteEditAbout.jsp</code>
+ * 
+ * @author Nischal Samji
+ *
+ */
+public class AboutTextValidator implements Validator {
+
+    @Override
+    public boolean supports(Class<?> arg0) {
+        ;
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/validator/AboutTextValidator.java]: ?

File: Quadriga/src/main/java/edu/asu/spring/quadriga/web/settings/WebsiteAboutEditController.java [hunk 4]
```diff
@@ -35,15 +42,30 @@
...
+        binder.setValidator(validator);
+    }
+
     @AccessPolicies({ @ElementAccessPolicy(type = CheckedElementType.PROJECT, paramIndex = 1, userRole = {
             RoleNames.ROLE_COLLABORATOR_OWNER, RoleNames.ROLE_PROJ_COLLABORATOR_ADMIN }) })
     @RequestMapping(value = "auth/workbench/projects/{ProjectId}/settings/editabout", method = RequestMethod.GET)
     public String editAbout(@PathVariable("ProjectId") String projectId, Model model, Principal principal)
             throws QuadrigaStorageException {
-        model.addAttribute("aboutText", aboutTextManager.getAboutTextByProjectId(projectId));
         IProject project = projectManager.getProjectDetails(projectId);
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/web/settings/WebsiteAboutEditController.java]: I believe there is an annotation to inject the project right into the method, isn't there? Then you can remove this line.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/web/settings/WebsiteAboutEditController.java [hunk 5]
```diff
@@ -58,15 +80,21 @@ public String editAbout(@PathVariable("ProjectId") String projectId, Model model
...
             RoleNames.ROLE_COLLABORATOR_OWNER, RoleNames.ROLE_PROJ_COLLABORATOR_ADMIN }) })
     @RequestMapping(value = "auth/workbench/projects/{ProjectId}/settings/saveabout", method = RequestMethod.POST)
-    public String saveAbout(@PathVariable("ProjectId") String projectId, @ModelAttribute("AboutTextBackingBean") AboutTextBackingBean formBean, Model model, Principal principal)
-            throws QuadrigaStorageException {
-        aboutTextManager.saveAbout(projectId, formBean.getTitle(), formBean.getDescription());
+    public ModelAndView saveAbout(@PathVariable("ProjectId") String projectId,
+            @Validated @ModelAttribute("aboutTextBean") AboutText formBean, BindingResult result, ModelAndView model,
+            Principal principal) throws QuadrigaStorageException {
+        model = new ModelAndView("auth/editabout");
         IProject project = projectManager.getProjectDetails(projectId);
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/web/settings/WebsiteAboutEditController.java]: same here. Just inject?

File: Quadriga/src/main/java/edu/asu/spring/quadriga/web/settings/WebsiteAboutEditController.java [hunk 6]
```diff
@@ -58,15 +80,21 @@ public String editAbout(@PathVariable("ProjectId") String projectId, Model model
...
-        model.addAttribute("project", project);
-        return "auth/editabout";
+        if (result.hasErrors()) {
+            model.addObject("aboutTextBean", formBean);
+            model.addObject("project", project);
+        } else {
+            aboutTextManager.saveAbout(projectId, formBean);
+            model.addObject("show_success_alert", true);
+            model.addObject("success_alert_msg", "You successfully edited the about text");
+            model.addObject("project", project);
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/web/settings/WebsiteAboutEditController.java]: duplicate code. move out of if block.

File: Quadriga/src/main/resources/locale/messages_en_US.properties [hunk 7]
```diff
@@ -215,6 +215,10 @@ account_approval.action.required=Please select "Approve" or "Deny".
 blog_title.required=Please enter title of blog entry.
 blog_description.required=Please enter a text for the new blog entry.
 
+# Settings Page error messages
+about_title.required=Please enter title for the about text.
```
Comment (jdamerow) [inline Quadriga/src/main/resources/locale/messages_en_US.properties]: Please enter a title for the about text.

File: Quadriga/src/test/java/edu/asu/spring/quadriga/service/publicwebsite/impl/AboutTextManagerTest.java [hunk 8]
```diff
@@ -57,20 +85,19 @@ public void testSaveAboutUpdateDTO() throws QuadrigaStorageException {
...
         argument = ArgumentCaptor.forClass(AboutTextDTO.class);
         Mockito.verify(mockedAboutTextDAO, Mockito.times(1)).saveOrUpdateDTO(argument.capture());
-        assertEquals("UniqueId2",argument.getValue().getId());
-        assertEquals("1st", argument.getValue().getProjectId());
-        assertEquals("NewTitle2", argument.getValue().getTitle());
-        assertEquals("NewDescription2", argument.getValue().getDescription());
+        assertEquals("UniqueId2", argument.getValue().getId());
+        assertEquals("PROJxyzwqr", argument.getValue().getProjectId());
+        assertEquals("UpdateTitle", argument.getValue().getTitle());
+        assertEquals("Update Description", argument.getValue().getDescription());
```
Comment (jdamerow) [inline Quadriga/src/test/java/edu/asu/spring/quadriga/service/publicwebsite/impl/AboutTextManagerTest.java:100]: These should be two methods. One that tests save and one that tests update. So that the test results give a clear indication what fails if something files (save might work, but update might fail).

File: Quadriga/src/test/java/edu/asu/spring/quadriga/service/publicwebsite/impl/AboutTextManagerTest.java [hunk 9]
```diff
@@ -95,13 +123,4 @@ public void testGetAboutTextByProjectId() throws QuadrigaStorageException {
         assertEquals("Test1", result.getId());
     }
 
-    @Test
-    public void testAboutTextNull() throws QuadrigaStorageException {
```
Comment (jdamerow) [inline Quadriga/src/test/java/edu/asu/spring/quadriga/service/publicwebsite/impl/AboutTextManagerTest.java:99]: why did you delete this test?

**Review by jdamerow — CHANGES_REQUESTED** [review]

File: Quadriga/src/main/java/edu/asu/spring/quadriga/validator/AboutTextValidator.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+     * not available error is thrown
+     * 
+     * @param obj
+     * @param err
+     */
+    @Override
+    public void validate(Object obj, Errors err) {
+        // validate all the input parameters
+        ValidationUtils.rejectIfEmptyOrWhitespace(err, "title", "about_title.required");
+        ValidationUtils.rejectIfEmptyOrWhitespace(err, "description", "about_description.required");
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/validator/AboutTextValidator.java:49]: you should also validate the submitted html only contains allowed tags and no javascript.

Comment (nischalsamji) [reply Quadriga/src/main/java/edu/asu/spring/quadriga/validator/AboutTextValidator.java:49]: If this validation is for enhancing security, can I simply escape the tag characters? I am not sure how I can validate only for javascript?

**Comment by nischalsamji** [comment]
Reopened pull request

File: Quadriga/src/main/java/edu/asu/spring/quadriga/service/publicwebsite/mapper/IAboutTextMapper.java (hunk 1 above)
Comment (nischalsamji) [reply Quadriga/src/main/java/edu/asu/spring/quadriga/service/publicwebsite/mapper/IAboutTextMapper.java]: changed the javadoc

File: Quadriga/src/main/java/edu/asu/spring/quadriga/validator/AboutTextValidator.java (hunk 2 above)
Comment (nischalsamji) [reply Quadriga/src/main/java/edu/asu/spring/quadriga/validator/AboutTextValidator.java]: moved the annotation

File: Quadriga/src/main/java/edu/asu/spring/quadriga/web/settings/WebsiteAboutEditController.java (hunk 6 above)
Comment (nischalsamji) [reply Quadriga/src/main/java/edu/asu/spring/quadriga/web/settings/WebsiteAboutEditController.java]: done.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/service/publicwebsite/mapper/IAboutTextMapper.java [hunk 11]
```diff
@@ -26,8 +26,8 @@
      * mapped {@link IAboutText} Object.
      * 
      * @param abtText
-     *            Domain Object that needs to be mapped to a
-     *            {@link AboutTextDTO} domain object.
+     *            DTO that needs to be mapped to a {@link AboutTextDTO} domain
+     *            object.
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/service/publicwebsite/mapper/IAboutTextMapper.java]: This is still wrong.

File: Quadriga/src/test/java/edu/asu/spring/quadriga/service/publicwebsite/impl/AboutTextManagerTest.java [hunk 12]
```diff
@@ -123,4 +127,12 @@ public void testGetAboutTextByProjectId() throws QuadrigaStorageException {
         assertEquals("Test1", result.getId());
     }
 
+    @Test(expected = NullPointerException.class)
+    public void testEmptyProject() throws QuadrigaStorageException {
+
+        Mockito.when(mockedAboutTextDAO.getDTOByProjectId("test24")).thenReturn(null);
+        aboutTextManager.getAboutTextByProjectId("test24");
```
Comment (jdamerow) [inline Quadriga/src/test/java/edu/asu/spring/quadriga/service/publicwebsite/impl/AboutTextManagerTest.java]: why would that throw a nullpointer? Nullpointers should rarely be thrown but appropriately handled. It's not unthinkable that this method might be called with a not existing id and it can't throw a nullpointer.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/validator/AboutTextValidator.java (hunk 10 above)
Comment (jdamerow) [reply Quadriga/src/main/java/edu/asu/spring/quadriga/validator/AboutTextValidator.java:49]: We talked about that. You haven't implemented that yet, right?

**Review by jdamerow — CHANGES_REQUESTED** [review]

**Comment by nischalsamji** [comment]
@jdamerow Addressed Review comments

File: Quadriga/src/test/java/edu/asu/spring/quadriga/service/publicwebsite/impl/AboutTextManagerTest.java [hunk 13]
```diff
@@ -127,11 +121,12 @@ public void testGetAboutTextByProjectId() throws QuadrigaStorageException {
...
     }
 
-    @Test(expected = NullPointerException.class)
+    @Test
     public void testEmptyProject() throws QuadrigaStorageException {
 
         Mockito.when(mockedAboutTextDAO.getDTOByProjectId("test24")).thenReturn(null);
-        aboutTextManager.getAboutTextByProjectId("test24");
+        IAboutText abtText = aboutTextManager.getAboutTextByProjectId("test24");
+        assertEquals(null, abtText);
```
Comment (jdamerow) [inline Quadriga/src/test/java/edu/asu/spring/quadriga/service/publicwebsite/impl/AboutTextManagerTest.java]: Isn't there an `assertNull` method?

File: Quadriga/src/test/java/edu/asu/spring/quadriga/validator/AboutTextValidatorTest.java [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+        emptyabtText = new AboutText();
+        emptyabtText.setId("TESTID");
+        emptyabtText.setProjectId("PROJtest");
+        emptyabtText.setDescription("");
+        emptyabtText.setTitle("");
+
+    }
+
+    @Test
+    public void testProperAboutText() {
```
Comment (jdamerow) [inline Quadriga/src/test/java/edu/asu/spring/quadriga/validator/AboutTextValidatorTest.java]: test method names should be `test_MethodToTestName_conditionTested`

**Comment by jdamerow** [comment]
ok to test

File: Quadriga/src/test/java/edu/asu/spring/quadriga/service/publicwebsite/impl/AboutTextManagerTest.java [hunk 15]
```diff
@@ -101,7 +102,7 @@ public void testSaveAboutUpdateDTO() throws QuadrigaStorageException {
     }
 
     @Test
-    public void testGetAboutTextByProjectId() throws QuadrigaStorageException {
+    public void getAboutTextByProjectId() throws QuadrigaStorageException {
```
Comment (jdamerow) [inline Quadriga/src/test/java/edu/asu/spring/quadriga/service/publicwebsite/impl/AboutTextManagerTest.java]: what about his one?

Comment (nischalsamji) [reply Quadriga/src/test/java/edu/asu/spring/quadriga/service/publicwebsite/impl/AboutTextManagerTest.java]: done

**Review by jdamerow — APPROVED** [review]
